Objects/weakrefobject.c

```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
     *list = newref;
 }
 
+/* See if we can reuse either the basic ref or proxy instead of creating a new
+ * weakref
+ */
+static PyWeakReference *
+try_reuse_basic_ref(PyWeakReference *ref, PyWeakReference *proxy,
+                    PyTypeObject *type, PyObject *callback)
+{
+    if (callback != NULL) {
+        return NULL;
+    }
+
+    PyWeakReference *cand = NULL;
+    if (type == &_PyWeakref_RefType) {
+        cand = ref;
+    }
+    if (type == &_PyWeakref_ProxyType) {
+        cand = proxy;
+    }
+
+#ifdef Py_GIL_DISABLED
+    if (cand != NULL && !_Py_TryIncref((PyObject **) &cand, (PyObject *) cand)) {
+        cand = NULL;
+    }
+#else
+    Py_XINCREF(cand);
+#endif
+
+    return cand;
+}
+
+static int
+is_basic_ref(PyWeakReference *ref)
+{
+    return (ref->wr_callback == NULL) && PyWeakref_CheckRefExact(ref);
+}
+
+static int
+is_basic_proxy(PyWeakReference *proxy)
+{
+    return (proxy->wr_callback == NULL) && PyWeakref_CheckProxy(proxy);
+}
+
+#ifdef Py_GIL_DISABLED
+
+static int
+is_basic_ref_or_proxy(PyWeakReference *wr)
+{
+    return is_basic_ref(wr) || is_basic_proxy(wr);
+}
+
+#endif
+
+/* Return the node that `newref` should be inserted after or NULL if `newref`
+ * should be inserted at the head of the list.
+ */
+static PyWeakReference *
+get_prev(PyWeakReference *newref, PyWeakReference *ref, PyWeakReference *proxy)
+{
+    if (is_basic_ref(newref)) {
+        return NULL;
+    }
+    if (is_basic_proxy(newref)) {
+        return ref;
+    }
+    return (proxy == NULL) ? ref : proxy;
+}
+
+typedef PyWeakReference *(*weakref_alloc_fn)(PyTypeObject *, PyObject *,
+                                             PyObject *);
+
+static PyWeakReference *
+get_or_create_weakref(PyTypeObject *type, weakref_alloc_fn allocate,
+                      PyObject *obj, PyObject *callback)
+{
+    if (!_PyType_SUPPORTS_WEAKREFS(Py_TYPE(obj))) {
+        PyErr_Format(PyExc_TypeError,
+                     "cannot create weak reference to '%s' object",
+                     Py_TYPE(obj)->tp_name);
+        return NULL;
+    }
+    if (callback == Py_None)
+        callback = NULL;
+
+    PyWeakReference **list = GET_WEAKREFS_LISTPTR(obj);
+    PyWeakReference *ref, *proxy;
+
+    LOCK_WEAKREFS(obj);
```

Comment:
This code structure seems more complicated than necessary because we lock twice. We only re-use weakreferences if they have no callback and one of three concrete types. Those three concrete types should be safe to allocate while holding the lock, even though it's not safe in general to allocate arbitrary types while holding the lock.

Let's restructure this so that we don't need the double checking. Something like:

1) First, check the type and callback to determine if we can reuse a weakref instance
2) If so, the with the lock held try to reuse an instance or allocate and insert a new one. Importantly the allocation and insertion happens without releasing the lock so we don't need to double check it.
3) Otherwise, allocate and return a new instance without acquiring the lock

Reply:
We don't reuse instances if `type` is a subtype of `weakref`. We only consider reuse if the type is *exactly* `_PyWeakref_RefType`, `_PyWeakref_ProxyType`, or `_PyWeakref_CallableProxyType`:

https://github.com/python/cpython/blob/29829b58a8328a7c2ccacaa74c1d7d120a5e5ca5/Objects/weakrefobject.c#L240-L246

If `type` is a subtype (i.e., not one of those three exact types), then we should allocate first (without the lock) and then insert unconditionally (with the lock).

Reply:
Oh yeah, of course. Sorry I failed to fully comprehend your original comment.